File: frontend/src/services/videoService.js
import apiService from './api';

class VideoService {
  /**
   * Generate a video from story and images
   * @param {Object} params - Video generation parameters
   * @param {string} params.story - Story content for the video
   * @param {Array} params.images - Array of image URLs
   * @param {string} [params.style="cinematic"] - Video style
   * @param {string} [params.duration="30"] - Video duration in seconds
   * @returns {Promise<Object>} Generated video data
   */
  async generateVideo({ story, images, style = "cinematic", duration = "30" }) {
    try {
      const response = await apiService.post('/api/video/generate', {
        story,
        images,
        style,
        duration
      });
      
      return response;
    } catch (error) {
      console.error('Video generation failed:', error);
      throw new Error(`Failed to generate video: ${error.message}`);
    }
  }

  /**
   * Upload a video to YouTube
   * @param {Object} params - YouTube upload parameters
   * @param {File} params.videoFile - Video file to upload
   * @param {string} params.title - Video title
   * @param {string} params.description - Video description
   * @param {Array} params.tags - Video tags
   * @param {string} [params.privacy="private"] - Privacy setting (private, unlisted, public)
   * @returns {Promise<Object>} Upload result
   */
  async uploadToYouTube({ videoFile, title, description, tags, privacy = "private" }) {
    try {
      const formData = new FormData();
      formData.append('video', videoFile);
      formData.append('title', title);
      formData.append('description', description);
      formData.append('tags', JSON.stringify(tags));
      formData.append('privacy', privacy);

      const response = await apiService.post('/api/video/youtube-upload', formData, {
        headers: {
          // Don't set Content-Type for FormData, let the browser set it
        }
      });
      
      return response;
    } catch (error) {
      console.error('YouTube upload failed:', error);
      throw new Error(`Failed to upload to YouTube: ${error.message}`);
    }
  }

  /**
   * Save a video project (if backend supports it)
   * @param {Object} video - Video object to save
   * @returns {Promise<Object>} Saved video data
   */
  async saveVideo(video) {
    try {
      const response = await apiService.post('/api/video/save', video);
      return response;
    } catch (error) {
      console.error('Video save failed:', error);
      throw new Error(`Failed to save video: ${error.message}`);
    }
  }

  /**
   * Get user's saved videos (if backend supports it)
   * @returns {Promise<Array>} Array of saved videos
   */
  async getSavedVideos() {
    try {
      const response = await apiService.get('/api/video/saved');
      return response;
    } catch (error) {
      console.error('Failed to fetch saved videos:', error);
      throw new Error(`Failed to fetch saved videos: ${error.message}`);
    }
  }

  /**
   * Delete a saved video (if backend supports it)
   * @param {string} videoId - ID of the video to delete
   * @returns {Promise<Object>} Deletion confirmation
   */
  async deleteVideo(videoId) {
    try {
      const response = await apiService.delete(`/api/video/${videoId}`);
      return response;
    } catch (error) {
      console.error('Video deletion failed:', error);
      throw new Error(`Failed to delete video: ${error.message}`);
    }
  }

  /**
   * Get video processing status
   * @param {string} videoId - ID of the video to check
   * @returns {Promise<Object>} Processing status
   */
  async getVideoStatus(videoId) {
    try {
      const response = await apiService.get(`/api/video/${videoId}/status`);
      return response;
    } catch (error) {
      console.error('Failed to get video status:', error);
      throw new Error(`Failed to get video status: ${error.message}`);
    }
  }

  /**
   * Download a video from URL
   * @param {string} videoUrl - URL of the video to download
   * @returns {Promise<Blob>} Video blob
   */
  async downloadVideo(videoUrl) {
    try {
      const response = await fetch(videoUrl);
      if (!response.ok) {
        throw new Error(`Failed to download video: ${response.status}`);
      }
      return await response.blob();
    } catch (error) {
      console.error('Video download failed:', error);
      throw new Error(`Failed to download video: ${error.message}`);
    }
  }
}

// Create and export a singleton instance
const videoService = new VideoService();
export default videoService;

File: backend/app/services/video_service.py
import os
import tempfile
import base64
import uuid
from typing import Optional, Dict, Any
from openai import OpenAI
import json

client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

async def generate_video_from_images(images: list, audio_base64: Optional[str] = None, duration: int = 10) -> str:
    """
    Generate a video from a sequence of images with optional audio
    
    Args:
        images: List of image URLs or base64 data
        audio_base64: Optional base64 encoded audio
        duration: Duration per image in seconds
    
    Returns:
        Base64 encoded video data
    """
    try:
        # This is a placeholder for video generation
        # In a real implementation, you would use a video processing library
        # like MoviePy, FFmpeg, or a cloud service
        
        # For now, we'll create a simple video file structure
        video_data = {
            "images": images,
            "audio": audio_base64,
            "duration": duration,
            "format": "mp4"
        }
        
        # Convert to base64 for storage
        video_json = json.dumps(video_data)
        video_base64 = base64.b64encode(video_json.encode()).decode()
        
        return video_base64
        
    except Exception as e:
        raise Exception(f"Video generation failed: {str(e)}")

async def save_video(video_base64: str, filename: str) -> str:
    """
    Save video to a file
    
    Args:
        video_base64: Base64 encoded video data
        filename: Name for the saved file
    
    Returns:
        Path to the saved file
    """
    try:
        # Create uploads directory if it doesn't exist
        upload_dir = "uploads/videos"
        os.makedirs(upload_dir, exist_ok=True)
        
        # Decode base64 video
        video_data = base64.b64decode(video_base64)
        
        # Save the file
        file_path = os.path.join(upload_dir, filename)
        with open(file_path, "wb") as f:
            f.write(video_data)
        
        return file_path
        
    except Exception as e:
        raise Exception(f"Failed to save video: {str(e)}")

async def upload_to_youtube(video_path: str, title: str, description: str = "", tags: list = None) -> Dict[str, Any]:
    """
    Upload video to YouTube (placeholder implementation)
    
    Args:
        video_path: Path to the video file
        title: Video title
        description: Video description
        tags: List of tags
    
    Returns:
        YouTube upload response
    """
    try:
        # This is a placeholder for YouTube upload
        # In a real implementation, you would use the YouTube Data API
        
        # Simulate upload process
        upload_id = str(uuid.uuid4())
        
        return {
            "success": True,
            "upload_id": upload_id,
            "youtube_url": f"https://youtube.com/watch?v={upload_id}",
            "title": title,
            "description": description,
            "tags": tags or []
        }
        
    except Exception as e:
        raise Exception(f"YouTube upload failed: {str(e)}")

async def get_video_status(video_id: str) -> Dict[str, Any]:
    """
    Get the status of a video processing job
    
    Args:
        video_id: The video ID
    
    Returns:
        Status information
    """
    try:
        # This would typically check a database or job queue
        # For now, return a mock status
        
        return {
            "video_id": video_id,
            "status": "completed",
            "progress": 100,
            "message": "Video processing completed successfully"
        }
        
    except Exception as e:
        raise Exception(f"Failed to get video status: {str(e)}")

async def process_video_upload(file_data: bytes, filename: str) -> str:
    """
    Process an uploaded video file
    
    Args:
        file_data: Raw file data
        filename: Original filename
    
    Returns:
        Path to the processed file
    """
    try:
        # Create uploads directory if it doesn't exist
        upload_dir = "uploads/videos"
        os.makedirs(upload_dir, exist_ok=True)
        
        # Generate unique filename
        file_ext = os.path.splitext(filename)[1]
        unique_filename = f"upload_{uuid.uuid4().hex[:8]}{file_ext}"
        file_path = os.path.join(upload_dir, unique_filename)
        
        # Save the uploaded file
        with open(file_path, "wb") as f:
            f.write(file_data)
        
        return file_path
        
    except Exception as e:
        raise Exception(f"Failed to process video upload: {str(e)}")

async def get_supported_video_formats() -> list:
    """
    Get list of supported video formats
    
    Returns:
        List of supported formats
    """
    return [
        {"format": "mp4", "description": "MPEG-4 video format"},
        {"format": "avi", "description": "Audio Video Interleave"},
        {"format": "mov", "description": "QuickTime Movie"},
        {"format": "wmv", "description": "Windows Media Video"},
        {"format": "flv", "description": "Flash Video"},
        {"format": "webm", "description": "WebM video format"}
    ]
